lib/Feature/FeatureModelParser.cpp

```diff
@@ -249,29 +250,242 @@ bool FeatureModelSxfmParser::parseFeatureTree(xmlChar *FeatureTree) {
   {
     std::stringstream Ss(reinterpret_cast<const char *>(FeatureTree));
     std::string To;
+    string Name;
+    bool Opt;
+    int LastIndentation = -1;
+    int OrGroupCounter = 0;
+    std::map<int, string> IndentationToParentMapping;
+
+    if (FeatureTree == nullptr) {
+      std::cerr << "Failed to read in feature tree. Is it empty?" << std::endl;
+      return false;
+    }
+
+    while (std::getline(Ss, To)) {
+      if (!std::getline(Ss, To)) {
+        return true;
+      }
+      Opt = false;
 
-    if (FeatureTree != nullptr)
-    {
-      while(std::getline(Ss, To)){
+      if (To.empty() || std::all_of(To.begin(),To.end(),isspace)) {
+        continue;
+      }
 
+      // For every line, count the indentation
+      // -1,0 or +1 additional indentations are allowed to the original one
+      // Otherwise, the format is violated
+      int CurrentIndentationLevel = countOccurrences(To, Indentation);
+      int Diff = LastIndentation - CurrentIndentationLevel;
+      if ((LastIndentation != -1) && (Diff < -1 | Diff > 1)) {
+        std::cerr << "Indentation error in feature tree." << std::endl;
+        return false;
       }
+
+      // Move pointer to first character after indentation
+      // The first character has to be a colon followed by the type of
+      // the feature (m for mandatory, o for optional, a for alternative)
+      std::string::size_type Pos = CurrentIndentationLevel * Indentation.length() + 2;
+      std::optional<std::tuple<string, string>> Cardinalities;
+      switch (To.at(Pos - 1)) {
+      case 'o':
+        // Code for optional
+        Opt = true;
+        break;
+      case 'g':
+        // Code for an or group with different cardinalities
+        Opt = false;
+        // Extract the cardinality
+        Cardinalities = extractCardinality(To);
+        // TODO: Process the cardinality
+        // Currently, we only accept cardinalities [1,1] (alternative group) and
+        // [1, *] (or group)
+
+        break;
+      case ' ':
+        // Code for alternative child
+        Pos--;
+        break;
+      }
+      // Extract the name
+      Name = readUntil(To, ' ', Pos + 1);
+
+      // Note that we ignore the ID and use the name of the feature
+      // as unique identifier.
+      if (Name.find_first_of('(') != std::string::npos) {
+        Name = readUntil(Name, '(', 0);
+      }
+
+      // Remove the cardinality
+      if (Name.find_first_of('[') != std::string::npos) {
+        Name = readUntil(Name, '[', 0);
+      }
+
+      // If there is no name, provide an artificial one
+      if (Name.empty()) {
+        OrGroupCounter++;
+        Name = "group_" + std::to_string(OrGroupCounter);
+      }
```

Comment:
after you checked the indentation you could just split the input line at whitespaces 

`std::vector<std::string> tokens;`
`split(To, " ", std::back_inserter(tokens));`